libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Memory/MemoryEventStore.cs

```diff
@@ -0,0 +1,133 @@
+using System.Collections.Concurrent;
+using EnergyOriginEventStore.EventStore.Serialization;
+using Newtonsoft.Json;
+
+namespace EnergyOriginEventStore.EventStore.Memory;
+
+public class MemoryEventStore : IEventStore
+{
+    private List<MessageEventArgs> _messages = new();
+    private ConcurrentQueue<Action> _actions = new();
+    [ThreadStatic] private bool isDraining = false;
+
+    public MemoryEventStore() { }
+
+    internal async Task Attach(MemoryEventConsumer consumer)
+    {
+        // NOTE: The following will enqueue an action that in turn will enqueue an action for each known message and then an action to begin listening for new messages.
+        // By letting a queued action perform the message deliveries and actual attaching, we avoid dealing with new messages during attachment.
+
+        _actions.Enqueue(async () =>
+        {
+            foreach (var message in _messages)
+            {
+                _actions.Enqueue(() =>
+                {
+                    consumer.OnMessage(this, message);
+                });
+            }
+
+            _actions.Enqueue(() =>
+            {
+                OnMessage += consumer.OnMessage;
+            });
+
+            await Drain().ConfigureAwait(false);
```

Comment:
Is this call to Drain() needed as it also on line 38?

Reply:
Yes, L38 starts the action that enqueues the real work, the L35 starts the real work actions that was just enqueued.